waku/v2/peermanager/topic_event_handler_test.go

```diff
@@ -0,0 +1,297 @@
+
+	// Subscribe to EventBus
+	err := pm.SubscribeToRelayEvtBus(relayEvtBus)
+	require.NoError(t, err)
+
+	// Register necessary protocols
+	pm.RegisterWakuProtocol(relay.WakuRelayID_v200, relay.WakuRelayENRField)
+	return pm, relayEvtBus
+}
+
+func TestSubscribeToRelayEvtBus(t *testing.T) {
+	log := utils.Logger()
+
+	// Host 1
+	r, h1, _ := makeWakuRelay(t, log)
+
+	// Host 1 used by peer manager
+	pm := NewPeerManager(10, 20, utils.Logger())
+	pm.SetHost(h1)
+
+	// Create a new relay event bus
+	relayEvtBus := r.Events()
+
+	// Subscribe to EventBus
+	err := pm.SubscribeToRelayEvtBus(relayEvtBus)
+	require.NoError(t, err)
+
+}
+
+func TestHandleRelayTopicSubscription(t *testing.T) {
+	log := utils.Logger()
+	pubSubTopic := "/waku/2/go/pm/test"
+	ctx := context.Background()
+
+	// Relay and Host
+	r, h1, _ := makeWakuRelay(t, log)
+	err := r.Start(ctx)
+	require.NoError(t, err)
+
+	// Peermanager with event bus
+	pm, _ := makePeerManagerWithEventBus(t, r, &h1)
+	pm.ctx = ctx
+	go pm.connectivityLoop(ctx)
+
+	// Subscribe to Pubsub topic
+	_, err = r.Subscribe(ctx, protocol.NewContentFilter(pubSubTopic))
+	require.NoError(t, err)
+
+	// Call the appropriate handler
+	select {
+	case e := <-pm.sub.Out():
+		switch e := e.(type) {
+		case relay.EvtRelaySubscribed:
+			{
+				eventDetails := (relay.EvtRelaySubscribed)(e)
+				pm.handleNewRelayTopicSubscription(eventDetails.Topic, eventDetails.TopicInst)
+			}
+		default:
+			require.Fail(t, "unexpected event arrived")
+		}
+
+	case <-ctx.Done():
+		require.Fail(t, "closed channel")
+	}
+
+	// Check Peer Manager knows about the topic
+	_, ok := pm.subRelayTopics[pubSubTopic]
+	require.True(t, ok)
+
+	// UnSubscribe from Pubsub topic
+	err = r.Unsubscribe(ctx, protocol.NewContentFilter(pubSubTopic))
+	require.NoError(t, err)
+
+	// Call the appropriate handler
+	select {
+	case e := <-pm.sub.Out():
+		switch e := e.(type) {
+		case relay.EvtRelayUnsubscribed:
+			{
+				eventDetails := (relay.EvtRelayUnsubscribed)(e)
+				pm.handleNewRelayTopicUnSubscription(eventDetails.Topic)
+			}
+		default:
+			require.Fail(t, "unexpected event arrived")
+		}
+
+	case <-ctx.Done():
+		require.Fail(t, "closed channel")
+	}
+
+	// Check the original topic was removed from Peer Manager
+	_, ok = pm.subRelayTopics[pubSubTopic]
+	require.False(t, ok)
+
+}
+
+func TestHandlePeerTopicEvent(t *testing.T) {
+	log := utils.Logger()
+	pubSubTopic := "/waku/2/go/pm/test"
+	ctx := context.Background()
+
+	hosts := make([]host.Host, 5)
+	relays := make([]*relay.WakuRelay, 5)
+
+	for i := 0; i < 5; i++ {
+		relays[i], hosts[i], _ = makeWakuRelay(t, log)
+		err := relays[i].Start(ctx)
+		require.NoError(t, err)
+	}
+
+	// Create peer manager instance with the first hosts
+	pm, eventBus := makePeerManagerWithEventBus(t, relays[0], &hosts[0])
+	pm.ctx = ctx
+	pm.RegisterWakuProtocol(relay.WakuRelayID_v200, relay.WakuRelayENRField)
+
+	// Connect host[0] with all other hosts to reach 4 connections
+	for i := 1; i < 5; i++ {
```

Comment:
Rather than adding peers this way, you can use PeerManager's AddPeer method which takes care of a bunch of these.

Reply:
I've simplified event handling test to focus on adding peers to topic and connectedness only. Getting to health  would require full stack and this could be tested better at interop level.